homeassistant/components/somfy/__init__.py

```diff
@@ -0,0 +1,160 @@
+"""
+Support for Somfy hubs.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/somfy/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant import config_entries
+from homeassistant.components.somfy import config_flow
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_TOKEN
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util import Throttle
+
+API = 'api'
+
+DEVICES = 'devices'
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
+
+DOMAIN = 'somfy'
+
+CONF_CLIENT_ID = 'client_id'
+CONF_CLIENT_SECRET = 'client_secret'
+
+SOMFY_AUTH_CALLBACK_PATH = '/auth/somfy/callback'
+SOMFY_AUTH_START = '/auth/somfy'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_CLIENT_ID): cv.string,
+        vol.Required(CONF_CLIENT_SECRET): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SOMFY_COMPONENTS = ['cover']
+
+
+async def async_setup(hass, config):
+    """Set up the Somfy component."""
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN] = {}
+
+    config_flow.register_flow_implementation(
+        hass, config[DOMAIN][CONF_CLIENT_ID],
+        config[DOMAIN][CONF_CLIENT_SECRET])
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={'source': config_entries.SOURCE_IMPORT},
+        ))
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up Somfy from a config entry."""
+    def token_saver(token):
+        _LOGGER.debug('Saving updated token')
+        entry.data[CONF_TOKEN] = token
+        hass.config_entries.async_update_entry(entry, data={**entry.data})
+
+    # Force token update.
+    from pymfy.api.somfy_api import SomfyApi
+    hass.data[DOMAIN][API] = SomfyApi(
+        entry.data['refresh_args']['client_id'],
+        entry.data['refresh_args']['client_secret'],
+        token=entry.data[CONF_TOKEN],
+        token_updater=token_saver
+    )
+
+    await update_all_devices(hass)
+
+    for component in SOMFY_COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component))
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Unload a config entry."""
+    if not hass.data[DOMAIN]:
+        hass.data.pop(DOMAIN)
```

Comment:
entry unload should only undo the work that was done as part of setting up the entry. setting `hass.data[DOMAIN]` is done by `async_setup` and should not be undone. It should be possible for an entry to be loaded, unloaded and loaded again.

Reply:
Thx for the explation. Code will be updated.